[Nathaniel Lukas] iP

dreammac3816547290 opened the pull request:
# Duke
>"The shorter way to do many things is to only do one thing at a time." - Mozart

Duke helps you keep track of ToDo, Event, and Deadline tasks. It's
- **TEXT-BASED**
- _simple_
- **_powerful_**

All you need to do is
1. download the duke.jar file from [here](https://github.com/dreammac3816547290/ip/releases).
2. open the command prompt and navigate to the location of the downloaded file.
3. type java -jar duke.jar and enter.
4. use it however you want! :smiley:

Features:
- [x] Add ToDos, Deadlines, and Events.
- [ ] List all tasks.
- [ ] Delete tasks.
- [ ] Mark tasks as done or not done.

If you are a Java programmer, you can use it to practice Java too. Here's the `main` method
```java
public static void main(String[] args) {
    new Duke().start();
}
```

dreammac3816547290 committed:
Skeletal version of Duke

dreammac3816547290 committed:
Add the ability to store text entered by the user

dreammac3816547290 committed:
Add the ability to mark tasks as done and change the status back to not done.

dreammac3816547290 committed:
Add support for tracking ToDos, Deadlines, and Events

dreammac3816547290 committed:
Use the input/output redirection technique to semi-automate the testing of Duke.

dreammac3816547290 committed:
Handle errors such as incorrect inputs entered by the user.

dreammac3816547290 committed:
Add support for deleting tasks from the list.

dreammac3816547290 committed:
Save the tasks in the hard disk automatically whenever the task list changes. Load the data from the hard disk when the project starts up.

dreammac3816547290 committed:
Store deadline and event dates as a java.time.LocalDate in the task objects. Accept dates in yyyy-mm-dd format and print in MMM dd yyyy format

dreammac3816547290 committed:
Merge branch 'branch-level-7'

dreammac3816547290 committed:
Merge branch 'branch-level-8'

lpohsien commented on src/main/java/Task.java:
```diff
@@ -0,0 +1,32 @@
+        isDone = false;
+    }
+
+    public String data() {
```
Perhaps the name for the method could be change to one that starts with a verb like `convertToData`?

lpohsien commented on src/main/java/Task.java:
```diff
@@ -0,0 +1,32 @@
+public class Task {
```
Perhaps this `Task` class and it's derived class could be put into a single package?

lpohsien commented on src/main/java/Task.java:
```diff
@@ -0,0 +1,32 @@
+        type = 'T';
+    }
+
+    public String getStatusIcon() {
```
Maybe you can add some method headers for the public methods?

lpohsien commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,143 @@
+                    System.out.println("Now you have " + tasks.size() + " tasks in the list.");
+                } else if (s.equals("deadline")) {
+                    String rest = sc.nextLine();
+                    String[] split = rest.split("/");
```
Considering that the `split` is an array of strings, perhaps it could named using the plural convention? Another way to name it could be `tokens` which might be more informative?

lpohsien reviewed (commented):
Overall, I found the code easy to follow. The coding standard was consistent with only a few nits that might need fixing. Descriptive headers for public classes and methods could be included to allow for better documentation.

Tan-Jin-Waye commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,143 @@
+            } catch (DukeException e) {
+                System.out.println(e.getMessage());
+            }
+        }
```
Do consider the use of a switch statement as opposed to a large if-else block.

~~~java
switch (s) {
case "bye":
    // Code for exit
    break;
case "list":
    // Code for list print
   break;
...

~~~

Tan-Jin-Waye commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,143 @@
+                tasks.add(new Event(split[2], split[3]));
+                if (split[1].equals("1")) {
+                    tasks.get(tasks.size() - 1).markAsDone();
+                }
```
Again, consider using a switch statement.
A switch fallthrough can be used to deal with the Deadline (line 115) and Event (line 120) cases elegantly.

~~~java
switch (split[0]) {
case "T":
    // Code for Todo
    break;
case "D":
   // Fallthrough
case "E":
   // Code for Deadline and Event
   break;
~~~

Tan-Jin-Waye commented on src/main/java/Task.java:
```diff
@@ -0,0 +1,32 @@
+public class Task {
+    protected String description;
+    protected boolean isDone;
+    protected char type;
```
Nice use of protected, though realistically not necessary for description and isDone.

Tan-Jin-Waye commented on src/main/java/Task.java:
```diff
@@ -0,0 +1,32 @@
+        return "[" + type + "][" + getStatusIcon() + "]" + description;
+    }
+
+    public void markAsDone() {
+        isDone = true;
+    }
+
+    public void markNotDone() {
+        isDone = false;
```
May want to include some code to check if the user is attempting to markAsDone an already done task, and same for markNotDone, and respond to user accordingly.

Tan-Jin-Waye reviewed (commented):
Code employed is sufficient to pass muster for Level 7/8, but better design choices need to be applied for later levels. I assume you're busy with other commitments at the moment, in which case all the best.

Jiayous.

dreammac3816547290 committed:
Extract Ui, Storage, Parser, TaskList, and Command into Classes

dreammac3816547290 committed:
Organize the classes into suitable java packages.

dreammac3816547290 committed:
Use Gradle to automate some of the build tasks of the project.

dreammac3816547290 committed:
Add JUnit tests to test the behavior of the code.

dreammac3816547290 committed:
Add JavaDoc comments to the code.

dreammac3816547290 committed:
Tweak the code to comply with a given coding standard

dreammac3816547290 committed:
Give users a way to find a task by searching for a keyword

dreammac3816547290 committed:
Resolved merge conflict between branch-A-JavaDoc and branch-A-CodingStandard

dreammac3816547290 committed:
Resolved merge conflict between branch-A-CodingStandard and branch-level-9

dreammac3816547290 committed:
Use checkStyle to detect coding style violations

dreammac3816547290 committed:
Add a GUI to Duke with JavaFX technology

dreammac3816547290 committed:
Use assert feature to document important assumptions that should hold at various points in the code
There are no checks to ensure the correctness of each aspect of the program (e.g. variable values at certain points in time).
It allows some bugs to be left undetected until an Exception happens and theprogram is terminated.
Adding assertions into the program acts as a sanity check and helps in ensuring the correctness of the program.

dreammac3816547290 committed:
Merge pull request #2 from dreammac3816547290/branch-A-Assertions

Use assert feature to document important assumptions that should hold…

dreammac3816547290 committed:
Examines the code and refactor to improve the code quality where necessary

Important exceptions were not thrown, causing the program to terminate when bugs occur.

Exceptions need to be thrown and catch to ensure the code works as intended and to eliminate bugs.

Important exceptions are thrown and catch, and the error message is displayed in Duke GUI, especially for exceptions regarding array out of bound and incorrect syntax.

dreammac3816547290 committed:
Provide a way for the user to find free times, the nearest date without task

dreammac3816547290 committed:
Merge pull request #3 from dreammac3816547290/branch-A-CodeQuality

Examines the code and refactor to improve the code quality where nece…

dreammac3816547290 committed:
Merge branch 'master' into branch-B-FindFreeTimes

dreammac3816547290 committed:
Merge pull request #4 from dreammac3816547290/branch-B-FindFreeTimes

Provide a way for the user to find free times, the nearest date witho…

dreammac3816547290 committed:
Fixed merge mistakes

dreammac3816547290 committed:
Add JUnit tests, to test nearly all code that can be tested automatically

dreammac3816547290 committed:
Add screenshot of Duke and add missing tag 'A-FullCommitMessage'

dreammac3816547290 committed:
Add a User Guide to the project in docs/README.md

dreammac3816547290 committed:
Replaced the README.md in the root directory

dreammac3816547290 committed:
Fixed image link in README.md

dreammac3816547290 committed:
Fixed build issue with gradle and modified Storage path
